chainer/training/updater.py

```diff
@@ -0,0 +1,366 @@
+            ignored and not used.
+
+    Attributes:
+        iteration: Current number of completed updates.
+
+    """
+    def __init__(self, iterator, optimizer, update_func=None, converter=None,
+                 device=None, loss_func=None):
+        if isinstance(iterator, iterator_module.Iterator):
+            iterator = {'main': iterator}
+        self._iterators = iterator
+
+        if isinstance(optimizer, optimizer_module.Optimizer):
+            optimizer = {'main': optimizer}
+        self._optimizers = optimizer
+
+        self._update_func = update_func or _default_update(
+            self, converter, device, loss_func)
+
+        self.iteration = 0
+
+    @property
+    def epoch(self):
+        return self._iterators['main'].epoch
+
+    @property
+    def epoch_detail(self):
+        return self._iterators['main'].epoch_detail
+
+    @property
+    def is_new_epoch(self):
+        return self._iterators['main'].is_new_epoch
+
+    def finalize(self):
+        for iterator in six.itervalues(self._iterators):
+            iterator.finalize()
+
+    def get_optimizer(self, name):
+        return self._optimizers[name]
+
+    def get_all_optimizers(self):
+        return dict(self._optimizers)
+
+    def get_iterator(self, name):
+        """Gets the dataset iterator of given name.
+
+        Args:
+            name (str): Name of the dataset iterator.
+
+        Returns:
+            ~chainer.dataset.Iterator: Corresponding dataset iterator.
+
+        """
+        return self._iterators[name]
+
+    def update(self):
+        self._update_func(self)
+        self.iteration += 1
+
+    def serialize(self, serializer):
+        for name, iterator in six.iteritems(self._iterators):
+            iterator.serialize(serializer['iterator:' + name])
+
+        for name, optimizer in six.iteritems(self._optimizers):
+            optimizer.serialize(serializer['optimizer:' + name])
+            optimizer.target.serialize(serializer['model:' + name])
+
+        self.iteration = serializer('iteration', self.iteration)
+
+
+def _default_update(updater, converter, device, loss_func):
+    if not converter:
+        converter = convert.concat_examples
+
+    iterator = updater.get_iterator('main')
+    optimizer = updater.get_optimizer('main')
+    loss_func = loss_func or optimizer.target
+
+    def update(_):
+        batch = iterator.next()
+        in_arrays = converter(batch, device)
+
+        if isinstance(in_arrays, tuple):
+            in_vars = tuple(variable.Variable(x) for x in in_arrays)
+            optimizer.update(loss_func, *in_vars)
+        elif isinstance(in_arrays, dict):
+            in_vars = {key: variable.Variable(x)
+                       for key, x in six.iteritems(in_arrays)}
+            optimizer.update(loss_func, **in_vars)
+        else:
+            in_var = variable.Variable(in_arrays)
+            optimizer.update(loss_func, in_var)
+
+    return update
+
+
+class ParallelUpdater(StandardUpdater):
+
```

Comment:
`PrallelUpdater` must have concrete implementations that are not implemented in base class. Namely, `finalize`, `get_optimizer`, `get_all_optimizers`, and `serialize`.